### PR TITLE
Loading CommandTimeout from Configuration, enhanced OptionsExtensions tests

### PR DESCRIPTION
**Pull request by mikary**
While working on #1168, I discovered that CommandTimeout settings were not being loaded from Configuration. This PR adds the missing functionality, and unit tests. I also reworked the unit tests for the ```SqlServerOptionsExtensions``` for consistency and more complete coverage of the configuration behaviors.
@lukewaters @ajcvickers

### CODE REVIEW

File: src/EntityFramework.SqlServer/SqlServerOptionsExtension.cs [hunk 1]
```diff
@@ -36,14 +40,14 @@ protected override void Configure(IReadOnlyDictionary<string, string> rawOptions
             base.Configure(rawOptions);
             if (!_maxBatchSize.HasValue)
             {
-                var MaxBatchSizeConfigurationKey = ProviderPrefix + ":" + MaxBatchSizeKey;
+                var maxBatchSizeConfigurationKey = ProviderPrefix + Constants.KeyDelimiter + MaxBatchSizeKey;
                 string maxBatchSizeString;
```
Comment (ajcvickers) [inline src/EntityFramework.SqlServer/SqlServerOptionsExtension.cs]: Why do we look for ProviderPrefix here, but not for the command timeout case?

File: src/EntityFramework.SqlServer/SqlServerOptionsExtension.cs [hunk 2]
```diff
@@ -36,14 +40,14 @@ protected override void Configure(IReadOnlyDictionary<string, string> rawOptions
...
+                if (rawOptions.TryGetValue(maxBatchSizeConfigurationKey, out maxBatchSizeString))
                 {
                     int maxBatchSizeInt;
                     if (!Int32.TryParse(maxBatchSizeString, out maxBatchSizeInt))
                     {
-                        throw new InvalidOperationException(Strings.IntegerConfigurationValueFormatError(MaxBatchSizeConfigurationKey, maxBatchSizeString));
+                        throw new InvalidOperationException(Strings.IntegerConfigurationValueFormatError(maxBatchSizeConfigurationKey, maxBatchSizeString));
                     }
                     _maxBatchSize = maxBatchSizeInt;
                 }
```
Comment (ajcvickers) [inline src/EntityFramework.SqlServer/SqlServerOptionsExtension.cs]: Can we reuse the parsing code from above--the two pieces of code seem almost identical.

**Comment by ajcvickers** [comment]
I am lost as to where we have ended up with MaxBatchSize and CommandTimeout. Can you summarize all the places each can be set, if and how they are provider-specific, and what we consider valid values and where we check for these?

**Comment by mikary** [comment]
Had a talk with @rowanmiller and @divega about the Relational vs Provider-Specific config settings. We are thinking that the ```CommandTimeout``` and ```MaxBatchSize``` should both be specified at the top (```EntityFramework:DbContext```) level in config but loaded in by provider-specific (```SqlServerOptionsExtension```) code.

For this change, we will put both public properties on the ```RelationalOptionsExtension``` but have the code for loading the settings on ```SqlServerOptionsExtension```. The ```ProviderPrefix``` will go away for now, but may be added again at a later date for provider-specific overrides on the relational settings.

**Comment by mikary** [comment]
Correction: The relational settings will be parsed and stored by the abstract class ```RelationalOptionsExtensions``` , and it will be up to the individual providers to use or ignore these settings on the provider-specific derived types.

File: src/EntityFramework.Relational/RelationalOptionsExtension.cs [hunk 3]
```diff
@@ -47,24 +50,54 @@ public virtual DbConnection Connection
         public virtual int? CommandTimeout
         {
             get { return _commandTimeout; }
+            [param: CanBeNull]
             set
             {
                 if (value.HasValue && value < 0)
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/RelationalOptionsExtension.cs]: Can command timeout be zero? If so, the exception message is wrong. If not, the check here is wrong.

File: src/EntityFramework.Relational/RelationalOptionsExtension.cs [hunk 4]
```diff
@@ -47,24 +50,54 @@ public virtual DbConnection Connection
...
                 {
-                    throw new ArgumentException(Strings.InvalidCommandTimeout);
+                    throw new InvalidOperationException(Strings.InvalidCommandTimeout);
```
Comment (ajcvickers) [inline src/EntityFramework.Relational/RelationalOptionsExtension.cs]: Why are these not ArgumentExceptions?

**Comment by ajcvickers** [comment]
:shipit:

**Comment by AndriySvyryd** [comment]
If max batch size is specified at the top-level wouldn't it make it impossible to specify different values for different providers? Since this is a perf tweaking setting it would be rare that the same value is optimal for different providers.

**Comment by mikary** [comment]
If we are using multiple providers for the same DbContext, it would not be possible to specify different values in Configuration unless we add provider specific overrides. Is it a common scenario to use the same context with multiple providers?

**Comment by AndriySvyryd** [comment]
I think it is common to have use the config file since it can be changed without recompiling and redeploying the app. It's fine as long as we maintain the ability to add a provider specific override.

**Comment by rowanmiller** [comment]
Yeah the plan we briefly discussed was having top level settings but then later adding support for adding provider specific sub-sections.

**Comment by rowanmiller** [comment]
BTW you currently can't swap providers from the config file, you have to change code to swap providers. Of course, you could have conditional code that swaps based on config settings... but in that case you can always include the different setting in the conditional code.

File: src/EntityFramework.Relational/RelationalOptionsExtension.cs (hunk 4 above)
Comment (mikary) [inline src/EntityFramework.Relational/RelationalOptionsExtension.cs]: To avoid duplicating the range checking logic, these exceptions will be thrown for an invalid setting from Configuration as well as setting the wrong value on the instance.
